lib/active_merchant/billing/gateways/omise.rb

```diff
@@ -0,0 +1,376 @@
+          response = parse(raw_response)
+        rescue ResponseError => e
+          raw_response = e.response.body
+          response = parse(raw_response)
+        rescue JSON::ParserError
+          response = json_error(raw_response)
+        end
+        response
+      end
+
+      def parse(body)
+        JSON.parse(body)
+      end
+
+      def json_error(raw_response)
+        {
+          message: 'Invalid response received from Omise API.  Please contact support@omise.co if you continue to receive this message.' +
+            '  (The raw response returned by the API was #{raw_response.inspect})'
+        }
+      end
+
+      def commit(method, endpoint, params=nil, options={})
+        response = https_request(method, endpoint, params, options)
+        Response.new(
+          successful?(response),
+          message_from(response),
+          response,
+          authorization: authorization_from(response),
+          test: test?,
+          error_code: successful?(response) ? nil : STANDARD_ERROR_CODE_MAPPING[error_code_from(response)]
+        )
+      end
+
+      def error_code_from(response)
+        (response['object'] == 'error') ? response['code'] : response['failure_code']
+      end
+
+      def message_from(response)
+        return 'Success' if successful?(response)
+        (response['message'] ? response['message'] : response['failure_message'])
+      end
+
+      def authorization_from(response)
+        response['id'] if successful?(response)
+      end
+
+      def successful?(response)
+        ( response.key?('object') and response['object'] != 'error' ) and response['failure_code'].nil?
+      end
+
+      def add_creditcard(post, creditcard)
+        return if creditcard.nil?
+        card = {
+          :number           => creditcard.number,
+          :name             => creditcard.name,
+          :security_code    => creditcard.verification_value,
+          :expiration_month => creditcard.month,
+          :expiration_year  => creditcard.year
+        }
+        post[:card] = card
+      end
+
+      def add_customer(post, options={})
+        return if post[:card].nil? or post[:card].match(/tokn(_test)?_[1-9a-z]+/)
+        post[:customer] = options[:customer_id] if post[:card].match(/card(_test)?_[1-9a-z]+/)
+      end
+
+      def add_customer_data(post, options={})
+        post[:description] = options[:description] if options[:description]
+        post[:email]       = options[:email] if options[:email]
+      end
+
+      def add_token_or_card(post, token_or_card, options={})
+        post[:card] = options[:token_id] || options[:card_id] || token_or_card
+      end
+
+      def add_amount(post, money, options)
+        post[:amount]      = amount(money)
+        post[:currency]    = options[:currency] || currency(money)
+        post[:description] = options[:description] if options.key?(:description)
+      end
+
+    end #OmiseGateway
+
+    class OmisePaymentToken < PaymentToken
```

Comment:
Also, FWIW, I wouldn't use the Stripe adapter as an example of good ActiveMerchant practice. It's kind of a bloated mess these days. I'd look at gateways like NMI, eWay Rapid, and Balanced (even though it's going away, :cry:).